### PR TITLE
[WB-7940] wandb.init() and singleton + do not repeat termwarm if requested WANDB_DIR/root_dir is not writable

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-7940

Description
-----------
In this PR:
- When you call, for example, `wandb.login()` (or `wandb.setup()`) before `wandb.init()`, the singleton will already be set up and so if e.g. env vars are set/changed in between, they will be ignored, which might not be what the user expects. 
When running `wandb.init()` we will now first check if the singleton exists and if so, check if the environment vars have changed and notify the user if they did.
image

- Prevent the terminal output from exploding if requested path (`WANDB_DIR`/`root_dir`) is not writable:
image

Testing
-------
Added a `yea` test.

### CODE REVIEW

**Review by vanpelt — APPROVED** [review]
🔥

File: functional_tests/login/01-timeout.yea [hunk 1]
```diff
@@ -0,0 +1,8 @@
+id: 0.login.01-timeout
+plugin:
+  - wandb
+tag:
+  skips:
+    - platform: win
```
Comment (kptkin) [inline functional_tests/login/01-timeout.yea:6]: just curious why?

File: functional_tests/login/02-login-init.yea [hunk 2]
```diff
@@ -0,0 +1,11 @@
+id: 0.login.02-login-init
+plugin:
+  - wandb
+tag:
+  skips:
+    - platform: win
```
Comment (kptkin) [inline functional_tests/login/02-login-init.yea:6]: same... as previous

File: functional_tests/login/01-timeout.yea (hunk 1 above)
Comment (dmitryduev) [reply functional_tests/login/01-timeout.yea:6]: heh, I don't know actually, it was there already, I just move it to a separate yea file :)

File: wandb/sdk/wandb_setup.py [hunk 3]
```diff
@@ -272,7 +272,7 @@ class _WandbSetup:
 
     _instance = None
 
-    def __init__(self, settings=None):
+    def __init__(self, settings=None) -> None:
```
Comment (kptkin) [inline wandb/sdk/wandb_setup.py:275]: nit: do you want to type settings?

File: wandb/sdk/wandb_setup.py [hunk 4]
```diff
@@ -283,7 +283,7 @@ def __getattr__(self, name):
         return getattr(self._instance, name)
 
 
-def _setup(settings=None, _reset=None):
+def _setup(settings=None, _reset: bool = False) -> Optional["_WandbSetup"]:
```
Comment (kptkin) [inline wandb/sdk/wandb_setup.py:286]: same

File: wandb/sdk/wandb_setup.py [hunk 5]
```diff
@@ -295,8 +295,9 @@ def _setup(settings=None, _reset=None):
     return wl
 
 
-def setup(settings=None):
+def setup(settings=None) -> Optional["_WandbSetup"]:
     ret = _setup(settings=settings)
+    # wandb.termlog("`wandb` session started.", repeat=False)
```
Comment (kptkin) [inline wandb/sdk/wandb_setup.py:300]: do you want to remove commented out code?

File: wandb/sdk/wandb_setup.py [hunk 6]
```diff
@@ -295,8 +295,9 @@ def _setup(settings=None, _reset=None):
     return wl
 
 
-def setup(settings=None):
+def setup(settings=None) -> Optional["_WandbSetup"]:
```
Comment (kptkin) [inline wandb/sdk/wandb_setup.py:298]: same do you want to type this?

**Review by kptkin — APPROVED** [review]
Awesome! Such a pleasure to review your PRs!!!

File: wandb/sdk/wandb_setup.py (hunk 6 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_setup.py:298]: This file is still not type-annotated :( I would leave it for a future PR, we should organize a type-annotation sprint or something.

File: wandb/sdk/wandb_setup.py (hunk 5 above)
Comment (dmitryduev) [inline wandb/sdk/wandb_setup.py:300]:
```suggestion
```

File: wandb/sdk/wandb_setup.py (hunk 6 above)
Comment (kptkin) [reply wandb/sdk/wandb_setup.py:298]: yeah that's why i nitted it haha

**Pull request merged**